src/dali_model_instance.cc

```diff
@@ -149,37 +205,66 @@ InputsInfo DaliModelInstance::GenerateInputs(const std::vector<TritonRequest>& r
     auto& request = requests[ri];
     ENFORCE(request.InputCount() == input_cnt,
             "Each request must provide the same number of inputs.");
-    for (uint32_t input_idx = 0; input_idx < input_cnt; ++input_idx) {
-      auto input = request.InputByIdx(input_idx);
-      auto input_byte_size = input.ByteSize();
-      auto input_buffer_count = input.BufferCount();
-      auto meta = input.Meta();
-      auto& idescr = input_map[meta.name];
-      for (uint32_t buffer_idx = 0; buffer_idx < input_buffer_count; ++buffer_idx) {
-        auto buffer = input.GetBuffer(buffer_idx, device_type_t::CPU, GetDaliDeviceId());
-        idescr.buffers.push_back(buffer);
+    auto idescrs = GenerateInputs(request);
+    if (ri == 0) {
+      for (auto &input: idescrs) {
+        input_map[input.meta.name] = input;
       }
-      if (idescr.meta.shape.num_samples() == 0) {
-        idescr.meta = meta;
-      } else {
-        ENFORCE(idescr.meta.type == meta.type,
-                make_string("Mismatched type for input ", idescr.meta.name));
-        idescr.meta.shape.append(meta.shape);
-      }
-      if (input_idx == 0) {
-        reqs_batch_sizes[ri] = meta.shape.num_samples();
-      } else {
-        ENFORCE(meta.shape.num_samples() == reqs_batch_sizes[ri],
-                "Each input in a request must have the same batch size.");
+    } else {
+      for (auto &input: idescrs) {
+        auto idescr = input_map.find(input.meta.name);
+        ENFORCE(idescr != input_map.end(), "Got unexpected input with name " + input.meta.name);
+        idescr->second.append(input);
       }
     }
+    reqs_batch_sizes[ri] = idescrs[0].meta.shape.num_samples();
   }
   for (const auto& descrs : input_map) {
     inputs.push_back(descrs.second);
   }
   return {inputs, reqs_batch_sizes};
 }
 
+std::vector<IDescr> DaliModelInstance::GenerateInputs(const TritonRequest &request) {
+  std::vector<IDescr> inputs(request.InputCount());
+  int num_samples = 0;
+  for (uint32_t input_idx = 0; input_idx < request.InputCount(); ++input_idx) {
+    auto input = request.InputByIdx(input_idx);
+    auto input_byte_size = input.ByteSize();
+    auto input_buffer_count = input.BufferCount();
+    auto meta = input.Meta();
+    auto &idescr = inputs[input_idx];
+    for (uint32_t buffer_idx = 0; buffer_idx < input_buffer_count; ++buffer_idx) {
+      auto buffer = input.GetBuffer(buffer_idx, device_type_t::CPU, GetDaliDeviceId());
+      idescr.buffers.push_back(buffer);
```

Comment:
Don't know how heavy that object is, but maybe consider std::move ?

Reply:
Good point. I added move constructors to some of the classes and minimized copying